tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,681 @@
+            }
+
+            // Are there enough permits remaining?
+            if (curr as u16) < num_permits {
+                return Err(TryAcquireError::NoPermits);
+            }
+
+            let next = curr - num_permits as usize;
+
+            match self.permits.compare_exchange(curr, next, AcqRel, Acquire) {
+                Ok(_) => return Ok(()),
+                Err(actual) => curr = actual,
+            }
+        }
+    }
+
+    fn poll_acquire(
+        &self,
+        cx: &mut Context<'_>,
+        needed: u16,
+        node: Pin<&mut Waiter>,
+        queued: bool,
+    ) -> Poll<Result<(), AcquireError>> {
+        let mut acquired = 0;
+
+        // If we are already in the wait queue, we need to lock the queue so we
+        // can access the wait queue entry's current state.
+        let (needed, mut lock) = if queued {
+            let lock = self.waiters.lock().unwrap();
+            // Safety: since we have acquired the lock, it is safe to look at
+            // the waiter's state.
+            let needed = node.state.with(|curr| unsafe { *curr });
+            (needed, Some(lock))
+        } else {
+            (needed as usize, None)
+        };
+
+        // First, try to take the requested number of permits from the
+        // semaphore.
+        let mut curr = self.permits.load(Acquire);
+        let mut waiters = loop {
+            // Has the semaphore closed?
+            if curr & CLOSED > 0 {
+                return Ready(Err(AcquireError::closed()));
+            }
+
+            let mut remaining = 0;
+            let (next, acq) = if curr + acquired >= needed {
+                let next = curr - (needed - acquired);
+                (next, needed)
+            } else {
+                remaining = (needed - acquired) - curr;
+                (0, curr)
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue.
+                if let Ok(l) = self.waiters.try_lock() {
+                    // If we were able to acquire the lock *now* without
+                    // blocking, we don't need to reload the semaphore's permit state.
+                    lock = Some(l)
+                } else {
+                    lock = Some(self.waiters.lock().unwrap());
+                    // While we were waiting to lock the wait list, additional
+                    // permits may have been released. Therefore, we will acquire a
+                    // new snapshot of the current state of the semaphore before
+                    // actually enqueueing the waiter..
+                    curr = self.permits.load(Acquire);
+                    continue;
+                }
+            }
+
+            match self
+                .permits
+                .compare_exchange_weak(curr, next, AcqRel, Acquire)
```

Comment:
did `compare_exchange_weak` have material impact? IME, `compare_exchange_weak` only comes w/ perf improvements when a failure does not require looping.

Reply:
I'd heard it should be used when the loop body is short, but that might not be the case (my understanding --- though I may be wrong --- is that it should generate the same instruction as `compare_exchange` on Intel, and I don't have access to an ARM machine for testing on...).

Reply:
I wouldn't worry about arm and i don't think loom models `compare_exchange_weak` yet. IIRC, it just calls `compare_exchange`, so i would suggest just sticking w/ `compare_exchange`.